src/main/scala/datalog/execution/ir/IROp.scala

```diff
@@ -190,6 +192,27 @@ case class InsertOp(rId: RelationId, db: DB, knowledge: KNOWLEDGE, override val
     }
 }
 
+case class ScanDiscoveredOp(rId: RelationId)(using JITOptions) extends IROp[EDB] {
+  val code: OpCode = OpCode.SCAN_DISCOVERED
+  override def run(storageManager: StorageManager): EDB =
+    storageManager.getDiscoveredEDBs(rId)
+  override def run_continuation(storageManager: StorageManager, opFns: Seq[CompiledFn[EDB]]): EDB =
+    run(storageManager) // leaf node, so no difference for continuation or run
+}
+
+case class NegateOp(arity: Int, override val children:IROp[EDB]*)(using JITOptions) extends IROp[EDB] {
```

Comment:
This may need to be renamed, since I'm not sure why an IRNode of type "Negate Operation" would be needed. Could we get away with storing if an atom is negated inside the joinIndexes and then when code is generated it checks if something is negated before calling the required storage manager methods? I'm not convinced we need an entire operation encoded in the IRTree to support negated atoms